core/src/main/java/com/linecorp/armeria/client/pool/DefaultKeyedChannelPool.java

```diff
@@ -280,20 +305,30 @@ protected boolean offerChannel(K key, Channel channel) {
 
     @Override
     public void close() {
-        for (Iterator<Deque<Channel>> i = pool.values().iterator(); i.hasNext();) {
-            final Deque<Channel> queue = i.next();
-            i.remove();
+        if (eventLoop.inEventLoop()) {
+            // While we'd prefer to block until the pool is actually closed, we cannot block for the channels to
+            // close if it was called from the event loop or we would deadlock. In practice, it's rare to call
+            // close from an event loop thread, and not a main thread.
+            doClose(false);
+        } else {
+            eventLoop.submit(() -> doClose(true)).syncUninterruptibly();
+        }
+    }
 
-            for (;;) {
-                final Channel ch = queue.pollFirst();
-                if (ch == null) {
-                    break;
-                }
+    private void doClose(boolean blocking) {
+        closed = true;
 
-                if (ch.isOpen()) {
-                    ch.close();
-                }
+        List<ChannelFuture> closeFutures = new ArrayList<>();
+
+        for (Channel ch : allChannels) {
+            if (ch.isOpen()) {
+                closeFutures.add(ch.close());
             }
         }
+        allChannels.clear();
+
+        if (blocking) {
+            closeFutures.forEach(ChannelFuture::syncUninterruptibly);
+        }
```

Comment:
Shouldn't we clear `closeFutures` as well?

Reply:
Since it's just a local variable that gets GC-able right away, figured it's better not to explicitly clear, but let me know if you'd still prefer it.

Reply:
Ah, yeah. No worries. Just missed that part. :-)